snuba/datasets/storage.py

```diff
@@ -167,7 +167,7 @@ def __init__(
         stream_loader: KafkaStreamLoader,
         query_splitters: Optional[Sequence[QuerySplitStrategy]] = None,
         mandatory_condition_checkers: Optional[Sequence[ConditionChecker]] = None,
-        replacer_processor: Optional[ReplacerProcessor] = None,
+        replacer_processor: Optional[ReplacerProcessor[Any]] = None,
```

Comment:
From https://github.com/getsentry/snuba/pull/1861#discussion_r631553844

Why don't you make `R = TypeVar("R", bound=Replacement)` covariant (like here: https://github.com/getsentry/snuba/blob/master/snuba/query/matchers.py#L17). By default generic types are invariant, so you are right, `ReplacerProcessor[Replacement]` is not subclass of `ReplacerProcessor[BaseReplacement]`. But here we are always returning a ReplacerProcessor, while we are never taking it in as a parameter, so we can should be able to declare that a ReplacerProcessor of a subtype of ReplacementBase is itself a subtype of ReplacerProcessor[Replacer].

Reply:
Wait thoug, what I said above may be nonsense.
The ReplacerProcessor both has methods that take in input `R` and methods that returns it `R`. 
Now snuba/replacer.py takes objects of type `R` from the processor and passes them back to the processor. If you make it covariant, the first part (getting objects of a subtype of R) would not be an issue, but the second part may be unless mypy is able to understand that they are the same object type.

I think it is worth trying but it may not work.

Reply:
I didn't know this existed, will try.

Reply:
mypy does not allow any use of covariant type params in method arguments: https://github.com/python/mypy/issues/7049

fairly heavy-handed...